src/test/java/com/google/devtools/build/lib/syntax/EvaluationTest.java

```diff
@@ -507,14 +507,14 @@ public void testListComprehensionDefinitionOrder() throws Exception {
     // because it doesn't record the scope in the syntax tree.)
     ev.new Scenario()
         .testIfErrorContains(
-            "local variable 'y' is referenced before assignment", //
+            "name 'y' is not defined", //
```

Comment:
The Starlark spec is unclear here, but our intention is that the previous behavior was correct. Pythons 2 and 3, for reasons I don't understand, resolve "for vars in expr" clauses (except the first) from left to right. That is, "for x in x" is permitted, because the binding of var x occurs before the resolution of body expression x.   (Possibly the reason is to allow loop-carried dependencies  on the previous value, but that seems awfully subtle.)

If I was designing from scratch, I would do what you have done here, which is to resolve each for-body before binding the for-vars.

Consequently, "foo for x in x" should be a dynamic error: local var x referenced before defined.

The outermost for clause is special (more rational): its body is evaluated in the environment outside the comprehension, and this outside the scope of its for-vars.

We could change the spec to make things more consistent, but we should discuss that outside the context of a code change.

Let's restrict our focus in this PR to fixing just the bug in question without reordering the resolver.